lightning/src/util/sweep.rs

```diff
@@ -0,0 +1,800 @@
+				*self = Self::PendingFirstConfirmation {
+					first_broadcast_hash: cur_hash,
+					latest_broadcast_height: cur_height,
+					latest_spending_tx,
+				};
+			},
+			Self::PendingFirstConfirmation { first_broadcast_hash, .. } => {
+				*self = Self::PendingFirstConfirmation {
+					first_broadcast_hash: *first_broadcast_hash,
+					latest_broadcast_height: cur_height,
+					latest_spending_tx,
+				};
+			},
+			Self::PendingThresholdConfirmations { .. } => {
+				debug_assert!(false, "We should never rebroadcast confirmed transactions.");
+			},
+		}
+	}
+
+	fn confirmed(
+		&mut self, confirmation_hash: BlockHash, confirmation_height: u32,
+		latest_spending_tx: Transaction,
+	) {
+		match self {
+			Self::PendingInitialBroadcast => {
+				// Generally we can't see any of our transactions confirmed if they haven't been
+				// broadcasted yet. Any transactions of the counterparty may confirm on-chain, but
+				// we hold the state lock after creating the initial entry until we tried
+				// broadcasting, so this should never be reachable via `transactions_confirmed`.
+				debug_assert!(false, "We should never confirm when we haven't broadcasted.");
+				*self = Self::PendingThresholdConfirmations {
+					first_broadcast_hash: confirmation_hash,
+					latest_broadcast_height: confirmation_height,
+					latest_spending_tx,
+					confirmation_height,
+					confirmation_hash,
+				};
+			},
+			Self::PendingFirstConfirmation {
+				first_broadcast_hash,
+				latest_broadcast_height,
+				..
+			} => {
+				debug_assert!(confirmation_height >= *latest_broadcast_height);
+				*self = Self::PendingThresholdConfirmations {
+					first_broadcast_hash: *first_broadcast_hash,
+					latest_broadcast_height: *latest_broadcast_height,
+					latest_spending_tx,
+					confirmation_height,
+					confirmation_hash,
+				};
+			},
+			Self::PendingThresholdConfirmations {
+				first_broadcast_hash,
+				latest_broadcast_height,
+				..
+			} => {
+				*self = Self::PendingThresholdConfirmations {
+					first_broadcast_hash: *first_broadcast_hash,
+					latest_broadcast_height: *latest_broadcast_height,
+					latest_spending_tx,
+					confirmation_height,
+					confirmation_hash,
+				};
+			},
+		}
+	}
+
+	fn unconfirmed(&mut self) {
+		match self {
+			Self::PendingInitialBroadcast => {
+				debug_assert!(
+					false,
+					"We should only mark a spend as unconfirmed if it used to be confirmed."
+				);
+			},
+			Self::PendingFirstConfirmation { .. } => {
+				debug_assert!(
+					false,
+					"We should only mark a spend as unconfirmed if it used to be confirmed."
+				);
+			},
+			Self::PendingThresholdConfirmations {
+				first_broadcast_hash,
+				latest_broadcast_height,
+				latest_spending_tx,
+				..
+			} => {
+				*self = Self::PendingFirstConfirmation {
+					first_broadcast_hash: *first_broadcast_hash,
+					latest_broadcast_height: *latest_broadcast_height,
+					latest_spending_tx: latest_spending_tx.clone(),
```

Comment:
Too bad we need to clone here. Could we take `self` by value and re-assign at the call site? Or maybe we'd need to introduce a wrapper around the state. Hrmm... In `OnionMessageRecipient`, we use `core::mem::swap` to avoid a clone. Possibly could work here.

Reply:
> I don't think taking `self` works here if `OutputSpendStatus` is ultimately owned by the `SweeperState`. Not sure what kind of wrapper you're suggesting?

Just a struct wrapping the enum. If the variants are never examined externally, then there is no need for the enum to be exposed. Then the wrapper can deconstruct the enum during state transitions.

Reply:
I guess this would still require taking `self` by value or having some uninitialized state to swap to before deconstructing, so maybe not worth it.